main/acle.md

```diff
@@ -4574,6 +4579,406 @@ two pointers, ignoring the tags.
+
+A program is [ill-formed](#ill-formed) if a function definition F
+uses `__arm_new` for some state that F also [shares](#shares-state)
+with its caller. For example:
+
+``` c
+  __arm_new("za") void f1() __arm_in("za") { ... }     // Invalid
+  __arm_new("za") void f2() __arm_out("za") { ... }    // Invalid
+  __arm_new("za") void f3() __arm_inout("za") { ... }  // Invalid
+  __arm_new("za") void f4() __arm_inout("zt0") { ... } // OK
+  __arm_new("zt0") void f5() __arm_inout("za") { ... } // OK
+```
+
+## Inline assembly
+
+Some ACLE implementations might support the GNU inline asm extension.
+For implementations that do, suppose that an inline asm occurs in a
+function F that [uses state](#uses-state) S. There are then two
+possibilities:
+
+*   S is not an input to the asm and S is not an output from the asm.
+    The contents of S immediately after executing the asm's instructions
+    are the same as the contents of S immediately before executing the
+    instructions. This is the default assumption.
+
+*   S might be an input to the asm, might be an output from the asm,
+    and might be changed by the asm's instructions. This is indicated by
+    adding S's string to the asm's clobber list. For example, if an asm's
+    clobber list includes `"za"`, ZA might be an input to the asm,
+    an output from the asm, or something that the asm changes.
+
+If instead the inline asm occurs in a function that does not [use
+state](#uses-state) S, the two possibilities are:
+
+*   S is not an input to the asm and S is not an output from the asm.
+    The contents of S immediately after executing the asm's instructions
+    are the same as the contents of S immediately before executing the
+    instructions. This is the default assumption.
+
+*   S is not an input to the asm and S is not an output from the asm.
+    However, the contents of S after executing the asm's instructions might
+    be different from the contents of S before executing the instructions.
+    This is indicated by adding S's string to the asm's clobber list.
+
+    If an asm takes this option for state that is controlled by PSTATE.ZA,
+    the asm itself is responsible for handling the [[AAPCS64]](#AAPCS64)
+    lazy save scheme, as described below.
+
+[[AAPCS64]](#AAPCS64) defines three possible states for ZA:
+“off”, “dormant” or “active”. These states describe the values
+of two things:
+
+*   the PSTATE.ZA bit
+
+*   the TPIDR2_EL0 register
+
+ZA is guaranteed to be active on entry to an inline asm A in a
+function F if at least one of the following is true:
+
+*   F [uses `"za"`](#uses-state) and A's clobber list includes `"za"`.
+
+*   F [uses `"zt0"`](#uses-state) and A's clobber list includes `"zt0"`.
+
+Otherwise, ZA can be in any state on entry to A if at least one of the
+following is true:
+
+*   F [uses](#uses-state) `"za"`
+
+*   F [uses](#uses-state) `"zt0"`
+
+Otherwise, ZA can be off or dormant on entry to A, as for what AAPCS64
+calls “private-ZA” functions.
```

Comment:
Can this be phrased the other way around:

* If F uses state S and A's clobber list includes S, then ZA is guaranteed to be active on entry to A.
* If F uses state S and A's clobber list does not include S, then ZA can be in any state on entry to A.
* If F does not use state S, then ZA can be off or dormant on entry to A.

?

Reply:
The difficulty with breaking it down like this is that the active/dormant/off states depend on both ZA and ZT0 together.  So:

* If F uses state S and A's clobber list does not include S, then ZA can be in any state on entry to A.
* If F does not use state S, then ZA can be off or dormant on entry to A.

is not true for S==ZA if F uses ZT0 and A's clobber list includes ZT0.  In particular, the last point would contradict the first point, since the first point would say ZA must be active (for S==ZT0), whereas the last point would say it must dormant or off (for S==ZA).

So I think it needs to be a chain of if-then-elses that deals with all PSTATE.ZA-based state together.

I can simplify the middle one to:

> Otherwise, ZA can be in any state on entry to A if F [uses](#uses-state) `"za"` or `"zt0"`

if you think that's better.  I wrote it the current way partly for consistency with the first “if”, partly to mimic the rules-based language in the new ISA supplements, and partly to avoid ambiguity about whether “or” includes “and” (which would become harder to deal with if there were ever more than two pieces of state controlled by PSTATE.ZA).

Reply:
Ah fair point. In that case I'm happy with the way it's written, I don't think combining those two bullet points into one makes a big difference, and at least it's now consistent with the case above.